File: LeetCode/1493/golang/lc1493.go
package lc1493

// Runtime complexity: O(n)
// Auxiliary space: O(1)
// Subjective level: Medium
func longestSubarray(nums []int) int {
	l := 0
	numZeroes := 0
	for _, num := range nums {
		if num == 0 {
			numZeroes += 1
		}
		if numZeroes > 1 {
			if nums[l] == 0 {
				numZeroes -= 1
			}
			l += 1
		}
	}
	return len(nums) - l - 1
}

File: LeetCode/1493/cpp/lc1493.cpp
#include <cassert>
#include <vector>

class Solution {
public:
  int longestSubarray(std::vector<int>& nums) {
    int ans = 0;
    int numZeroes = 0;
    int rs = 0; // range start
    int re = 0; // range end
    while (re < nums.size()) {
      if (nums[re] == 0) {
        ++numZeroes;
      }
      while (numZeroes == 2) {
        if (nums[rs] == 0) {
          --numZeroes;
        }
        ++rs;
      }

      if (re - rs > ans) {
        ans = re - rs;
      }
      ++re;
    }
    return ans;
  }
};

int main() {
  Solution s;
  {
    std::vector<int> query = {1, 1, 0, 1};
    assert(s.longestSubarray(query) == 3);
  }
  {
    std::vector<int> query = {0, 1, 1, 1, 0, 1, 1, 0, 1};
    assert(s.longestSubarray(query) == 5);
  }
  {
    std::vector<int> query = {1, 1, 1};
    assert(s.longestSubarray(query) == 2);
  }
}
